### PR TITLE
DynamoDb Enhanced perf testing

### PR DESCRIPTION
**Pull request by dagnir**
## Description
Tests overhead of GET and PUT item over the low level client, as well as
performance of those operations compared to the V1 mapper.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by dagnir** [comment]
```
Benchmark                                       (testItem)   Mode  Cnt      Score    Error  Units
EnhancedClientPutOverheadBenchmark.enhPutHuge          N/A  thrpt    8   6178.310 ± 21.855  ops/s
EnhancedClientPutOverheadBenchmark.enhPutSmall         N/A  thrpt    8  21834.816 ± 62.910  ops/s
EnhancedClientPutOverheadBenchmark.enhPutTiny          N/A  thrpt    8  24995.407 ± 49.975  ops/s
EnhancedClientPutOverheadBenchmark.llPut              TINY  thrpt    8  25125.061 ± 71.506  ops/s
EnhancedClientPutOverheadBenchmark.llPut             SMALL  thrpt    8  23217.554 ± 89.430  ops/s
EnhancedClientPutOverheadBenchmark.llPut              HUGE  thrpt    8   8534.843 ± 25.766  ops/s
```

**Comment by dagnir** [comment]
```
Benchmark                                       (testItem)   Mode  Cnt      Score     Error  Units
EnhancedClientGetOverheadBenchmark.enhGet             TINY  thrpt    8  21985.693 ±  70.213  ops/s
EnhancedClientGetOverheadBenchmark.enhGet            SMALL  thrpt    8  18645.558 ± 129.404  ops/s
EnhancedClientGetOverheadBenchmark.enhGet             HUGE  thrpt    8   4751.360 ±  12.167  ops/s
EnhancedClientGetOverheadBenchmark.llGet              TINY  thrpt    8  23480.263 ±  79.170  ops/s
EnhancedClientGetOverheadBenchmark.llGet             SMALL  thrpt    8  21336.071 ±  63.168  ops/s
EnhancedClientGetOverheadBenchmark.llGet              HUGE  thrpt    8   6619.252 ±  20.321  ops/s
EnhancedClientPutOverheadBenchmark.enhPutHuge          N/A  thrpt    8   5905.032 ±  25.948  ops/s
EnhancedClientPutOverheadBenchmark.enhPutSmall         N/A  thrpt    8  20660.983 ±  63.780  ops/s
EnhancedClientPutOverheadBenchmark.enhPutTiny          N/A  thrpt    8  23994.051 ±  74.239  ops/s
EnhancedClientPutOverheadBenchmark.llPut              TINY  thrpt    8  24335.511 ±  88.384  ops/s
EnhancedClientPutOverheadBenchmark.llPut             SMALL  thrpt    8  21922.118 ±  31.407  ops/s
EnhancedClientPutOverheadBenchmark.llPut              HUGE  thrpt    8   8064.792 ±  22.985  ops/s

```

File: test/sdk-benchmarks/pom.xml [hunk 1]
```diff
@@ -134,6 +134,16 @@
             <artifactId>netty-nio-client</artifactId>
             <version>${awsjavasdk.version}</version>
         </dependency>
+        <dependency>
+            <groupId>software.amazon.awssdk</groupId>
+            <artifactId>dynamodb-enhanced</artifactId>
+            <version>${awsjavasdk.version}-PREVIEW</version>
```
Comment (spfink) [inline test/sdk-benchmarks/pom.xml]: Add a TODO here so we don't forget?

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/EnhancedClientGetOverheadBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+@BenchmarkMode(Mode.Throughput)
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+@Fork(2)
+@State(Scope.Benchmark)
+public class EnhancedClientGetOverheadBenchmark {
+    private static final AwsJsonProtocolFactory JSON_PROTOCOL_FACTORY = AwsJsonProtocolFactory
+            .builder()
+            .clientConfiguration(SdkClientConfiguration.builder()
+                    .option(ENDPOINT, URI.create("https://dynamodb.amazonaws.com"))
```
Comment (spfink) [inline test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/EnhancedClientGetOverheadBenchmark.java]: What's this endpoint?

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/EnhancedClientGetOverheadBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    private static final PutItemRequestMarshaller PUT_ITEM_REQUEST_MARSHALLER =
+            new PutItemRequestMarshaller(JSON_PROTOCOL_FACTORY);
+
+    private static final V2ItemFactory ITEM_FACTORY = new V2ItemFactory();
+
+    private final Key testKey = Key.builder().partitionValue("key").build();
+
+
+    @Benchmark
+    public Object llGet(TestState s) {
```
Comment (spfink) [inline test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/EnhancedClientGetOverheadBenchmark.java]: I think we can write out better method names here

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/ItemFactory.java [hunk 4]
```diff
@@ -0,0 +1,922 @@
...
+
+    public final HugeBean hugeBean() {
+        HugeBean b = new HugeBean();
+        b.setHashKey(randomS());
+        b.setStringAttr(randomS());
+        b.setBinaryAttr(randomBytes());
+        b.setListAttr(Arrays.asList(randomS(),
+                randomS(),
+                randomS(),
+
```
Comment (spfink) [inline test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/ItemFactory.java]: Why the seemingly random new lines? Also can we just for loop something here?

File: test/sdk-benchmarks/pom.xml [hunk 5]
```diff
@@ -134,6 +134,16 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-dynamodb</artifactId>
+            <version>1.11.748</version>
```
Comment (spfink) [inline test/sdk-benchmarks/pom.xml]: Do we really need v1 here long term? Once we have a good baseline for v2, can we just track v2?

File: test/sdk-benchmarks/pom.xml (hunk 1 above)
Comment (dagnir) [reply test/sdk-benchmarks/pom.xml]: Sure

File: test/sdk-benchmarks/pom.xml (hunk 5 above)
Comment (dagnir) [reply test/sdk-benchmarks/pom.xml]: We can remove it in the future, but while we're iterating on the performance we should have this in here.

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/ItemFactory.java (hunk 4 above)
Comment (dagnir) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/ItemFactory.java]: The lines break them up in to powers of 2. I can change to using a loop

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/EnhancedClientGetOverheadBenchmark.java (hunk 2 above)
Comment (dagnir) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/enhanced/dynamodb/EnhancedClientGetOverheadBenchmark.java]: This was to stop the numerous profile loader related log messages

File: buildspecs/benchmarks.yml [hunk 6]
```diff
@@ -10,4 +10,4 @@ phases:
       - mvn install -P quick -pl :sdk-benchmarks --am
       - mvn install -pl :bom-internal
       - cd test/sdk-benchmarks
-      - mvn exec:exec
\ No newline at end of file
+      - java -jar target/benchmarks.jar -f 1 -wi 8 -i 8 software.amazon.awssdk.benchmark.enhanced.dynamodb
```
Comment (zoewangg) [inline buildspecs/benchmarks.yml]: intended change?

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/BenchmarkRunner.java [hunk 7]
```diff
@@ -74,6 +83,7 @@ public static void main(String... args) throws RunnerException, JsonProcessingEx
         benchmarksToRun.addAll(ASYNC_BENCHMARKS);
         benchmarksToRun.addAll(PROTOCOL_BENCHMARKS);
         benchmarksToRun.addAll(COLD_START_BENCHMARKS);
+        benchmarksToRun.addAll(MAPPER_BENCHMARKS);
```
Comment (zoewangg) [inline test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/BenchmarkRunner.java]: Should we add ddb enhanced client data to `baseline.json` for regression detection?

File: buildspecs/benchmarks.yml (hunk 6 above)
Comment (dagnir) [reply buildspecs/benchmarks.yml]: just while I'm testing these changes in codebuild

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/BenchmarkRunner.java (hunk 7 above)
Comment (dagnir) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/BenchmarkRunner.java]: See previous comment for getting baseline

**Pull request merged**